Fixed run issues.

FarisHijazi opened the pull request:
There were many issues that prevented running the tool from source, most of these were command line issues as well as referencing variables that didn't exist.

I have resolved these issues and now there are no runtime errors.

FarisHijazi committed:
update .gitignore to ignore pychache

FarisHijazi committed:
fix: cli.py: importing irrelevant commands, filtered commands: ['Train', 'WhatIs']

the fifty.commands variable included many members that were not needed and this was causing issues. These members were being included because all the imports were included in train.py and whatis.py.

Example of such members: keras.layers.Conv1D
These were being executed and errors were popping up

FarisHijazi committed:
fix #2: train.py unresolved names and variables

- fixed accessing of variables
- options['<input>'] was also fixed by replacing it with options['--data-dir']
- df = pd.DataFrame, dtype was added to ensure proper typing
- removed global variables (around line 214)
- NOTE: still unsure about what to do with `self.new_model` around line 293

FarisHijazi committed:
updated default values for opt parser, and extended usage

also extended usage for `train` command. Some options like `-f` are needed but haven't been added! So I added those

FarisHijazi committed:
fix: cli.py: fixed bug where __version__ was not getting imported

also now you can run the cli.py file since `if __name__=='__main__'` was added

FarisHijazi committed:
refactor: train.py code cleanup

FarisHijazi committed:
fixed self.new_model variable, replaced it with self.model_name

FarisHijazi committed:
refactor: moved code to build_model() function

- also in get_best(), created the dictionary on spot (cleaner code)
- also added double quotes to some print messages

FarisHijazi committed:
doc: Updated README.md, added example command line

FarisHijazi committed:
fix: train.py: "Value error: Input arrays should have the same number of..."

Error: "Value error: Input arrays should have the same number of samples as target arrays."

this was caused because when the data was formatted, the same variables were used, this should not be done, also made sure that trimmed data would never be 0

FarisHijazi committed:
fix: train.py: "Value error: Input arrays should have the same number of..."

Error: "Value error: Input arrays should have the same number of samples as target arrays."

this was caused because when the data was formatted, the same variables were used, this should not be done, also made sure that trimmed data would never be 0

FarisHijazi committed:
refactor: moved build_model() from train.py to framework.py

also some code cleanup and simplification

FarisHijazi committed:
feat: now you can train the model even if output exists, the output hparams will be loaded

the output/parameters.csv file will be loaded if it exists, and if it exists then there's no need for the hparam exploration, so this saves a lot of time and work can be continued across sessions
- fix: also now loading saved model will check the output directory

FarisHijazi committed:
fix: input arguments, no there's a default value for "<input>", also fixed self.args

self.args was referenced to `options`, which is not accurate, now both whatis.py and train.py follow the same naming for self.args and self.options

FarisHijazi committed:
feat: added --paramspace option which now reads params from json

- refactor: moved hparamspace exploration to its own function: explore_hparam_space()
- also added some options to cli.py

FarisHijazi committed:
fix: whatis.py bugs (sys module was not imported)

FarisHijazi committed:
doc: added IEEE paper pdf, and ignored output files

FarisHijazi committed:
Merge branch 'master' of https://github.com/FarisHijazi/fifty

FarisHijazi committed:
small fixes + refactor: dataframe now is assigned dictionary values

- '--params' was parsed as bool (fixed)
- no when loading data, '--force' is not set, then the data will be loaded, and '--force' will delete the entire tree (just like in the original repo)
- line 219: self.df is now assigned the parameters, instead of assigning the list. This is a problem because python dictionaries are NOT ordered, so it was shear luck that it was working before, now it is more robust
- fixed mismatch columns while training, this was happening because when the dataframe is saved, it saves the index too
-

FarisHijazi committed:
feat: unified model naming, continue prev model training, --epochs option

- now train_network has an option to load the previous model if it was found
- unified naming: the model_dir() function makes safe filenames based on the hparams json, this allows generating those names consistently across session
- --epochs option will control how many epochs the final model will be trained (this is still subject to early stopping)

FarisHijazi committed:
tweaks/fixes: shuffling data, turning percent from int to float, fix .json

- parameter_space.json had trailing commas which is not allowed, be ware of this

FarisHijazi committed:
feat: added tensorboard logging

FarisHijazi committed:
feat: added tensorboard logging

mittalgovind reviewed (changes requested):
Hello Faris,

Thank you for showing interest in our work. Thank you for all the minor changes suggested by your commits. I would like to accept most of the changes except that the paper PDF should be accessible through the arXiv link only.


Also, I plan to add the following features so as to make the tool easier to use.
- A better plotting function for displaying the memory layouts in a more understandable and readable way.
- Easier way to swap dataset components. Current FFT-75 - scenario #'1 dataset has PDF fragments that contain JPGs in them. This confuses the classifier making the accuracy suffer. With this feature partial or complete replacement of one fragment with new fragments should be done automatically. E.g., replacing all or 30% of the fragments of PDFs with new fragments of PDFs.
- prevent handling of TensorFlow variables explicitly, i.e., making the tool cross-compatible.

Let me know if you are interested in doing it.

FarisHijazi committed:
fix: refactor: build_model() now can accept optimizer and loss args

and cleanup
-  replaced Train.build_model() with build_model()
- now tensorboard logs will not be separated by session (each session will not be in a timestamp folder on it's own)
- improved loading in train_network()

FarisHijazi committed:
fix: convention: now empty parameters are represented with 0 instead of None

paramameters.json will take 0 for empty parameters instead of null

FarisHijazi committed:
refactor: organize code, print statements and imports

FarisHijazi committed:
feat: added command to diagram, using Graphviz

Note that graphviz must be installed for this to work, and it is (optional)

FarisHijazi committed:
refactor: little code fixes and comments, also removed tf warning

also created function framework.py::get_latest_tfevent_summary()

FarisHijazi committed:
feat: adding support for autoencoder, adding "enc_dim" parameter keyword

NOTE: now the model cannot be built ahead of time, and thus the model summary cannot be displayed before fit().

This is something that I need help with, but using `model.build(input_shape=(input_length,))` is not working

FarisHijazi committed:
feat: train.py no longer accepts scenario option, no_of_classes is autodetected

no_of_classes is detected automatically from the training dataset (depending on unique values from the target values)

FarisHijazi committed:
feat: created framework::build_autoencoder()


fix: train.py::build_autoencoder()

the layers were not connected correctly at all, the decoder got the decoder's input instead of the decoder's output

FarisHijazi committed:
refactor: train.py catch use-case for force=False and dataset doesn't exist + reorganizing

When --force is not set, the code will try to load the dataset, but if the dataset isn't found then, then an exception should be raised (which is what happens now)

- reorganizing if-statements and improving printing
- code cleanup: now  parameters['accuracy']=0, parameters['loss']=np.inf, before the run, so we don't have to keep setting them, we can just break and they'll have the correct values
- also made code safer when adding the new row to the dataframe by making sure it has the same keys (line:284)
- added another exception to whitelist

FarisHijazi committed:
feat: added --autoencoder option to train the autoencoder

- validation loss is used as a monitor for the autoencoder (since accuracy can't be used)
- there is some work to do in terms of the filenaming ( model_dir() parameters)

FarisHijazi committed:
feat: added "loss" to dataframe, now get_best() uses loss in case accuracy is all the same

- also now get_best() raises an exception if the dataframe is empty
- fixed accessing the dataframe using .loc[]

refactor: train.py::get_best() prints the best accuracy/loss

FarisHijazi committed:
chore(gitignore): created _ignore/ directory and ignored it

also ignored output_autoencoder/

FarisHijazi committed:
chore: created requirements.txt

use pip freeze requirements.txt to install the packages

FarisHijazi committed:
feat: added a default hparamspace_autoencoder file, default is autoselected

- added -t option, shorthand for --autoencoder
- renamed parameter_space.json to hparamspace.json
- default value is autoselected depending if --autoencoder is set or not

FarisHijazi committed:
refactor: feat: train.py moved checking of inputs to __init__

- now the dataframe will not carry irrelevent info like 'dense' (for the autoencoder doesn't have that), it will only carry info depending on the parameters json.
- also renamed enc_dim to use_encoder

FarisHijazi committed:
feat: added autoencoder evaluation after training + some refactoring

FarisHijazi committed:
feat: connect autoencoder to classifier model

- finally putting things together
- updated paramspace jsons
- added multigpu support for autoencoder

FarisHijazi committed:
refactor: code and print messages cleanup

FarisHijazi committed:
merged master and encoder branch

FarisHijazi committed:
fix: fixed import structure

import fifty.commands results in an error, replaced with import commands

FarisHijazi committed:
removed IEEE pdf

FarisHijazi committed:
doc: created tree.txt

FarisHijazi commented:
alright I'm fine with removing the PDF, but I'm not sure how to do it on GitHub, perhaps you could accept the pull request then revert that commit?

FarisHijazi commented:
also in terms of helping with the plotting, maybe we could workout a deal where you'd give me some credit in your paper perhapse?
I'll be happy to help.

But currently I'm just experimenting with the current repo and trying different stuff.
